File: src/ruby/lib/geodelta/id_util.rb
# encoding: utf-8

module GeoDelta
  module IdUtil
    def self.get_all_delta_ids(level)
      return (0..7).map { |id| [id] } if level == 1
      return (0..7).to_a.product(*([(0..3).to_a] * (level - 1)))
    end
  end
end

File: src/java/geodelta/src/jp/nayutaya/geodelta/IdUtil.java

package jp.nayutaya.geodelta;

public class IdUtil
{
    public static byte[][] getAllDeltaIds(final int level)
    {
        final int len = (int)Math.pow(2, (level * 2) + 1);
        final byte[][] ids = new byte[len][];
        for ( int i = 0; i < len; i++ )
        {
            final byte[] id = new byte[level];
            id[0] = (byte)((i >> (2 * (level - 1))) & 7);
            for ( int lv = 2; lv <= level; lv++ )
            {
                id[lv - 1] = (byte)((i >> (2 * (level - lv))) & 3);
            }
            ids[i] = id;
        }
        return ids;
    }
}
